# MNST Classification

### grid search for best Hyperparameters

In [2]:
from sklearn.datasets import load_digits
from sklearn.model_selection import GridSearchCV
from sklearn import svm 

In [3]:
import multiprocessing

In [4]:
cpus = multiprocessing.cpu_count()
print(cpus)

8


In [5]:
params = [
    {
        'kernel':['linear','rbf','poly','sigmoid'],
        'C':[0.1,0.2,0.5,1.0,2.0,5.0]
    }   
]


## Dataset Preparation 

In [7]:
data = load_digits()
X = data.data
Y = data.target
print(X.shape, Y.shape)

(1797, 64) (1797,)


## Logistic Regression

In [8]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score

In [11]:
lr = LogisticRegression()
lr.fit(X,Y)
cross_val_score(lr,X,Y,scoring="accuracy",cv=5).mean()

/home/harsh_linux/.local/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/home/harsh_linux/.local/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-

0.9137650882079852

#### Logistic Regression gives around 921% accuracy

In [13]:
from sklearn import svm

In [21]:
svc = svm.SVC()
svc.fit(X,Y)
cross_val_score(svc,X,Y,scoring="accuracy",cv=5).mean()

0.9632838130609718

## Right set of parameters 

 - Gridseach gives us the hyper parameters

In [25]:
gs = GridSearchCV(estimator=svm.SVC(), param_grid=params, scoring="accuracy", cv=5, n_jobs=cpus)

In [26]:
gs.fit(X, Y)

GridSearchCV(cv=5, estimator=SVC(), n_jobs=8,
             param_grid=[{'C': [0.1, 0.2, 0.5, 1.0, 2.0, 5.0],
                          'kernel': ['linear', 'rbf', 'poly', 'sigmoid']}],
             scoring='accuracy')

In [27]:
gs.best_estimator_

SVC(C=5.0)

In [28]:
gs.best_score_

0.9738502011761063

## SVM with Polynomial Kernel, C=0.1 gives 97% accuracy on the Dataset!